### PR TITLE
planner: support union statements with ctes

### PR DESCRIPTION
**Pull request by maxenglander**
## Description

Support `WITH ... SELECT ... UNION ... SELECT` CTE syntax.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/15311

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/sqlparser/ast_format.go [hunk 1]
```diff
@@ -74,6 +74,10 @@ func (node *CommentOnly) Format(buf *TrackedBuffer) {
 
 // Format formats the node.
 func (node *Union) Format(buf *TrackedBuffer) {
+	if node.With != nil {
+		buf.astPrintf(node, "%v", node.With)
```
Comment (dbussink) [inline go/vt/sqlparser/ast_format.go:78]: @maxenglander Can you also add a unit test for this to the `sqlparser` package?

**Review by systay — APPROVED** [review]

File: go/vt/sqlparser/ast_format.go (hunk 1 above)
Comment (systay) [reply go/vt/sqlparser/ast_format.go:78]: `go/vt/sqlparser/parse_test.go` is a good place for this

**Comment by GrahamCampbell** [comment]
Any chance we can treat this as a bug in the CTE implementation and backport to v19?

**Review by dbussink — APPROVED** [review]
Marked this as  a bug, since the parser always should roundtrip queries properly but that doesn't happen here.

**Pull request merged**